lib/vm/src/instance.rs

```diff
@@ -688,6 +688,189 @@ impl Instance {
     }
 }
 
+/// This is an intermediate type that manages the raw allocation and
+/// metadata when creating an [`Instance`].
+///
+/// This type will free the allocated memory if it's dropped before
+/// being used.
+///
+/// The [`UnpreparedInstance::instance_layout`] computes the correct
+/// layout to represent the wanted [`Instance`].
+///
+/// Then we use this layout to allocate an empty `Instance` properly.
+pub struct UnpreparedInstance {
+    /// The buffer that will contain the [`Instance`] and dynamic fields.
+    instance_ptr: NonNull<Instance>,
+    /// The layout of the `instance_ptr` buffer.
+    instance_layout: Layout,
+    /// Information about the offsets into the `instance_ptr` buffer for
+    /// the dynamic fields.
+    offsets: VMOffsets,
+    /// Whether or not this type has transferred ownership of the
+    /// `instance_ptr` buffer. If it has not when being dropped,
+    /// the buffer should be freed.
+    consumed: bool,
+}
+
+impl Drop for UnpreparedInstance {
+    fn drop(&mut self) {
+        if !self.consumed {
+            // If `consumed` has not been set, then we still have ownership
+            // over the buffer and must free it.
+            let instance_ptr = self.instance_ptr.as_ptr();
+            unsafe {
+                std::alloc::dealloc(instance_ptr as *mut u8, self.instance_layout);
```

Comment:
I believe we should also use `ptr::drop_in_place(instance_ptr);` in order to properly drop `Instance`. For example, if `Instance` has a `Drop` implementation, `drop_in_place` will call it, whilst `alloc::dealloc` won't.

Reply:
The actual code in this PR might also be unsafe. Nothing forbids `self.instance_ptr` to be dangling, and I don't know what happens when one tries to dealloc a dangling pointer 🤔.

Reply:
> I believe we should also use ptr::drop_in_place(instance_ptr); in order to properly drop Instance. For example, if Instance has a Drop implementation, drop_in_place will call it, whilst alloc::dealloc won't.


The intent of this type was that it would never exist with the `Instance` written to it, it was purely managing the buffer prior to that step.  Due to feedback from Syrus, an intermediate type which made the distinction of having the `Instance` written to it was removed and now there are a few lines where, if dropped, this type would do the wrong thing (however `consumed` is also set there, so it wouldn't free anything anyways).

I think it probably does not make sense to add drop logic for `Instance` here unless we end up merging types more somehow.  I still like having an additional type / passing it to into the constructor of `InstanceHandle`, it would allow us to reduce unsafety and simplify the code but I got strong feedback against it.

--------

> The actual code in this PR might also be unsafe. Nothing forbids self.instance_ptr to be dangling, and I don't know what happens when one tries to dealloc a dangling pointer 🤔.

That's a good thought but as far as I can tell it is not possible.  The only way it could be possible is because this type is defined in a huge file where people could access its field directly.  I think the solution is to move it into another file where its private fields can remain private.  Because we railroad the user with (now an extremely simple) type-state machine, the user can't use the API incorrectly as there's no way to give a pointer that we didn't create ourselves.

Reply:
All good to me :-). Thanks for your reply.